Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1100 @@
+      public var message: Message
+
+      /// Metadata received from the server at the end of the RPC.
+      public var trailingMetadata: Metadata
+    }
+
+    /// Whether the RPC was accepted or rejected.
+    ///
+    /// The `success` case indicates the RPC completed successfully with an 'ok'
+    /// status code. The `failure` case indicates that the RPC was rejected or
+    /// couldn't be completed successfully.
+    public var result: Result<Body, RPCError>
+
+    public init(result: Result<Body, RPCError>) {
+      // ...
+    }
+
+    // Note: it's possible to provide a number of conveneince APIs on top:
+
+    /// The metadata received from server at the start of the RPC.
+    ///
+    /// The metadata will be empty if `result` is `failure`.
+    public var metadata: Metadata {
+      get {
+        // ...
+      }
+    }
+
+    /// The message returned from the server.
+    ///
+    /// Throws if the RPC was rejected or failed.
+    public var message: Message {
+      get throws {
+        // ...
+      }
+    }
+
+    /// The metadata received from server at the end of the RPC.
+    public var trailingMetadata: Metadata {
+      get {
+        // ...
+      }
+    }
+  }
+
+  public struct Stream<Message: Sendable>: Sendable {
+    public struct Body {
+      /// Metadata received from the server at the start of the RPC.
+      public var metadata: Metadata
+
+      /// A sequence of messages received from the server ending with metadata
+      /// if the RPC succeeded.
+      ///
+      /// If the RPC fails then the sequence will throw an error.
+      public var bodyParts: any AsyncSequence<BodyPart>
+
+      public enum BodyPart: Sendable {
+        case message(Message)
+        case trailers(Metadata)
+      }
+    }
+
+    /// Whether the RPC was accepted or rejected.
+    ///
+    /// The `success` case indicates the RPC was accepted by the server for
+    /// processing, however, the RPC may still fail by throwing an error from its
+    /// `messages` sequence. The `failure` case indicates that the RPC was
+    /// rejected by the server.
+    public var result: Result<Body, RPCError>
+
+    public init(result: Result<Body, RPCError>) {
+      // ...
+    }
+
+    // Note: it's possible to provide a number of conveneince APIs on top:
+
+    /// The metadata received from server at the start of the RPC.
+    ///
+    /// The metadata will be empty if `result` is `failure`.
+    public var metadata: Metadata {
+      get {
+        // ...
+      }
+    }
+
+    /// The stream of messages received from the server.
+    public var messages: any AsyncSequence<Message> {
+      get {
+        // ...
+      }
+    }
+
+    /// The metadata received from server at the end of the RPC.
+    public var trailingMetadata: Metadata {
+      get {
+        // ...
+      }
+    }
+  }
+}
+
+// MARK: - Supporting types
+
+/// A sink for values which are produced over time.
+public protocol Writer<Element: Sendable>: Sendable {
```

Comment:
We probably want to support `~Copyable` types here and all the other protocols/generic types as well and have a conditional `Copyable` conformance. I don't think the latter is expressible yet but we have to keep it in mind.